# README #

BioKotlin leverages the performance and type safety of the JVM languages with ease of Python like scripting.  
Because the Kotlin language is compiled on the fly it can be thousands of times faster for certain applications than 
scripting languages.  BioKotlin is in the early stages of development, but we expect to support nucleotide and 
protein sequence manipulation, fast sequence IO, alignment, motif, pathway support.  Plus interoperability with
 [GATK](https://gatk.broadinstitute.org/hc/en-us), SamTools [HTSJDK](https://samtools.github.io/htsjdk/), 
 BioJava, [BioPython](https://biopython.org), [TASSEL](https://www.maizegenetics.net/tassel) for GWAS and Genome Wide Prediction, and the PHG - the 
 pangenome representation.
 
 We expect to have extensive cross language support with the use of GraalVM's support of JVM, Python, and R (FastR). 
 
 You can follow our current progress at our [Trello Board](https://bitbucket.org/bucklerlab/biokotlin/addon/trello/trello-board) 
 
### What is this repository for? ###

* This is the main repository for BioKotlin development
* We follow a Git-Flow branching structure
* Since we are in early development, most of the code is in the 
[develop](https://bitbucket.org/bucklerlab/biokotlin/branch/develop) and associated branches

### How do I get set up? ###
 
* Configuration for a User
    * JupyterLab with Kotlin
    
* Configuration for a Power User (or Developer)
    * IntelliJ with Kotlin support  
    * JupyterLab with Kotlin

* Dependencies
* 

* How to run tests
* Deployment instructions

### Contribution guidelines ###

* Writing tests
* Code review
* Other guidelines

### Who do I talk to? ###

* Edward Buckler - esb33@cornell.edu
* Terry Casstevens - tmc46@cornell.edu

### ToDo ###
1. Making the 
8. Interop between languages using GraalVM

