# README #

This is multilanguage platform for modeling molecular biology and physiology for applied breeding.

### What is this repository for? ###

* This project is an initial experiment to model the central dogma of molecular biology in a way 
that would be useful to make breeding predictions.
* Definitions of terms and the structure of the model
* How to run the model is different languages
* How to create a new function and share it.
* [Learn Markdown](https://bitbucket.org/tutorials/markdowndemo)

### How do I get set up? ###

* Data files - place test in data folder (do not sync with Git)
    * B73 Reference genome proteins and GFF
    * Flowering time/clock EggNog Orthologs
    * Kremling expression data for above genes and NAM founders
    * Mini-PHG for NAM founders and only the flowering genes
 
* Configuration
    * GraalVM
    * FastR
    * JupyterLab with Kotlin

* Dependencies
* Database configuration

* How to run tests
* Deployment instructions

### Contribution guidelines ###

* Writing tests
* Code review
* Other guidelines

### Who do I talk to? ###

* Edward Buckler - esb33@cornell.edu
* Other community or team contact

### ToDo ###
1. Define terms and flow (Markdown file in repository)
2. Define Parameters and Outputs
3. Test data
4. Model pseudocode - how would you like to run this pipeline in scientific English
5. Create basic functions and flow in Kotlin, R, and/or Python (can just move random values)
6. Test code for basic functions
7. Start implementing easiest of the real functions
8. Interop between languages using GraalVM

